### PR TITLE
Add has_more_values to chain filter endpoints

### LINKED ISSUE #16102: Filter dropdown with 10k characters of total distinct values sends query to data source every time dropdown is opened
**Describe the bug**
From https://github.com/metabase/metabase/pull/15943#pullrequestreview-654993149 - continuation of #13832
When a dashboard has a "A list of all values" filter, which has more than 10,000 characters of total distinct values (ex. `select length(string_agg(my_column, '')) from my_table`) are not present in `metabase_fieldvalues` (or if cached field values has never been scanned), then clicking a filter widget on a dashboard will make a query on data source for 100 results every time the dropdown is opened.

This happens even if [scanning is disabled](https://www.metabase.com/docs/latest/administration-guide/01-managing-databases.html#choose-when-metabase-syncs-and-scans) in Admin > Databases > (db) > Scheduling-tab > select "Never, I'll do this manually if I need to" and save.

**Workaround** Admin > Data Model > (db) > (table) > (column) :gear: > change the "Filtering on this field" from "A list of all values" to either "Search box" or "Plain input box"

If you have a lot of tables/fields, then it might be easier to make the change in the application database.
Go to the application database table `metabase_field` and change the column `has_field_values` to `none` or `search`

**To Reproduce**
Using Sample Dataset, but on MariaDB/Postgres, so it's easier to see queries being executed.
1. Admin > Data Model > Sample Dataset > People > City > set as "A list of all values" (dropdown)
2. Simple question > Sample Dataset > People - save as "Q1" to dashboard
3. On dashboard add a category filter and connect to "Q1".City - save dashboard
4. Refresh the dashboard and click filter widget to open - every click will generate a query like this to the data source:
```
SELECT `PEOPLE`.`CITY` AS `CITY` FROM `PEOPLE` GROUP BY `PEOPLE`.`CITY` ORDER BY `PEOPLE`.`CITY` ASC LIMIT 100
```
5. If the instance doesn't have scanned values (e.g. disabled or not scanned yet), then it should make the query to the data source first time, but then the results should be memorized (for disabled) or possibly stored in the appdb `metabase_fieldvalues` table (for not scanned yet), so subsequent requests are faster.

**Expected behavior**
Should use the appdb to get values, which is generally much faster than datasource. And should be memorized.

**Information about your Metabase Installation:**
Testing commit `aad24de` (almost 0.39.2)

Related to #16101 #16104

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by qnkhuat**
In #23200 we added the `has_more_values` flag to FieldValues so that FE can know when to do an in-browser search vs call BE to search for values of a Field.

But in that PR we only include `has_more_values` for `GET /api/field/:id/values` endpoint, which we used to get values of a field in QB.

In this PR I made the changes to include `has_more_values` for all of our `chain-filter` endpoints, and they are:
- GET /api/dashboard/:id/params/:param-id/values
- GET /api/dashboard/:id/params/:param-id/search
- GET /api/embed/:id/params/:param-id/values
- GET /api/embed/:id/params/:param-id/search
- GET /api/public/:id/params/:param-id/values
- GET /api/public/:id/params/:param-id/search

Currently, these APIs will return a list of values, after these changes they will return a map with shape:
```clojure
{:values          [1, 2 ,3]
 :has_more_values true}
```

### Note for reviewers
this PR seems to have a lot of change but most are updating existing tests for these endpoints to match the new shape.
So most tests file could be ignored, the only exception is the new tests I added at the end of `test/metabase/models/params/chain_filter_test.clj`

FE's PR to apply this change:  #24022
Epic #23162
Closes #16102

### CODE REVIEW

File: src/metabase/api/dashboard.clj [hunk 1]
```diff
@@ -604,14 +606,17 @@
        ;; TODO - we should combine these all into a single UNION ALL query against the data warehouse instead of doing a
        ;; separate query for each Field (for parameters that are mapped to more than one Field)
       (try
-        (let [results (distinct (mapcat (if (seq query)
-                                          #(chain-filter/chain-filter-search % constraints query :limit result-limit)
-                                          #(chain-filter/chain-filter % constraints :limit result-limit))
-                                        field-ids))]
+        (let [results (map (if (seq query)
```
Comment (qnkhuat) [inline src/metabase/api/dashboard.clj:609]: any Clojure magic trick to do this without building an immediate concatnated results?

File: src/metabase/models/params/chain_filter.clj [hunk 2]
```diff
@@ -621,9 +633,10 @@
   (or (when-let [unremapped-values (not-empty (matching-unremapped-values query v->human-readable))]
         (let [query-constraint  {field-id (set unremapped-values)}
               constraints       (merge constraints query-constraint)
-              values            (unremapped-chain-filter field-id constraints options)]
-          (add-human-readable-values values v->human-readable)))
-      []))
+              result            (unremapped-chain-filter field-id constraints options)]
+          (update result :values #(add-human-readable-values % v->human-readable))))
```
Comment (bshepherdson) [inline src/metabase/models/params/chain_filter.clj:637]: Same.

File: src/metabase/models/params/chain_filter.clj [hunk 3]
```diff
@@ -643,13 +656,14 @@
...
-        query  (str/lower-case query)]
-    (cond->> (filter (fn [s]
-                       (when s
-                         (str/includes? (str/lower-case s) query)))
-                     values)
-      limit (take limit))))
+  (let [{:keys [values has_more_values]} (cached-field-values field-id constraints nil)
+        query                            (str/lower-case query)]
+    {:values (cond->> (filter (fn [s]
+                                (when s
```
Comment (bshepherdson) [inline src/metabase/models/params/chain_filter.clj:662]: This `filter` might be clearer as `(for [... :when ...] ...)`?

**Review by bshepherdson — APPROVED** [review]

File: src/metabase/models/params/chain_filter.clj (hunk 3 above)
Comment (qnkhuat) [reply src/metabase/models/params/chain_filter.clj:662]: it does, but I feel like using `filter` has better sematic for `filtering` than using `for :when`.

so I'd keep using `filter` for this.

**Pull request merged**